Testing setup

gabriel-marinkovic opened the pull request:
Introduces a basic unit testing framework for fun.

Tests use a custom file format, explained & showcased in `test/test_framework.test.fun`. Tests can be run with `./fun -test`. Some features:
- Ability to define multiple independent tests in a single file.
- Ability to have a test pass when a compiler error, runtime error, or a specific compiler error message appears.
- Tests are isolated in separate processes and ran in parallel.
- Ability to re-run a test that failed with the exact same source code & compiler rng seed.
- Ability to specify rng seed for the compiler for each test.
- Basic stress test mode with `./fun -stress_test`, which runs all tests until one of them fails.
- Cute CLI.

Also started adding some basic tests:
- Copy-pasted old tests from `src_fun/test` into new system (they fail at this point though, something to investigate).
- Basic BigNum functionality
- Arithmetic
- Type sizes
- Basic variable usage
- Operator precedence
- Basic alias behavior
- Numeric literals

Also reworked the CLI a bit.
- Compile a file with `./fun <path>`
- Interpret an expression with `./fun -run <expr>`
- Run all tests with `./fun -test`
- Run the stress test mode with `./fun -stress_test`
- Run a specific failed testcase with `./fun -test_runner ...`. The exact command is created by the `-test` mode once a test fails.

gabriel-marinkovic committed:
initial test running mode. defined test file format, test execution with process separation, fail error reporting, re-running failed tests in main process, output formatting, stress test mode.

gabriel-marinkovic committed:
basic test_assert intrinsic

gabriel-marinkovic committed:
wrote some basic tests for math, variables, sizes of types, aliasing. added old tests to test framework. other minor fixes.

Kalinovcic commented on src_common/common_linux.inl:
```diff
@@ -1283,6 +1285,57 @@ void terminate_process_without_waiting(Process* process)
+        QPC start_time  = current_qpc();
+        do
+        {
+            if (waitpid(pid, &status, WNOHANG) > 0)
```
Check for EINTR, in case the call was interrupted by a signal handler. https://linux.die.net/man/7/signal

Kalinovcic replied:
Nevermind, seems WNOHANG guarantees EINTR is impossible.

Kalinovcic commented on src_common/common_linux.inl:
```diff
@@ -1283,6 +1285,57 @@ void terminate_process_without_waiting(Process* process)
+                return true;
+            }
+
+            usleep(10000); // sleep for 10ms
```
Do an exponential backoff, waiting `1 << min(attempts++, 15)` micros

Kalinovcic commented on src_fun/main.cpp:
```diff
@@ -3,46 +3,110 @@
+    if (argc > 1 && argv[1][0] == '-')
+        first_flag_arg = consume(wrap_string(argv[1]), 1);
+
+    if (first_flag_arg == "test_runner"_s)
```
Merge -test_runner and -test into one flag, such that:

* `fun -test` runs all tests
* `fun -test math-int-div0 math-float-div0` only runs those two
* `fun -test test/basic.test.fun` runs all from that file
* `fun -test math-*` runs all matching the `math-*` wildcard

This is necessary to make this usable once there are a lot of tests. I only want to run the ones for compiler interop for example, not all the adds and subtracts and hash tables, ...

gabriel-marinkovic replied:
`-test_runner` is just an entry point for the subprocess which runs the test. I will rename it to `-test_process` to make that more clear.

I agree with everything else.

gabriel-marinkovic replied:
> `fun -test test/basic.test.fun` runs all from that file

Actually not doing this, you can run all tests from this file by just doing `fun -test basic`.

Kalinovcic commented on src_fun/main.cpp:
```diff
@@ -3,46 +3,110 @@
+    if (first_flag_arg == "run"_s)
+    {
+        String_Concatenator code_cat = {};
+        add(&code_cat, "run unit { debug\n"_s);
```
Not placing parantheses around the user input is making me nervous. Make it

```
debug(
<arg> <arg>
);```

Kalinovcic commented on src_fun/main.cpp:
```diff
@@ -3,46 +3,110 @@
+        add(&code_cat, "run unit { debug\n"_s);
+        for (umm i = 2; i < argc; i++)
+        {
+            add(&code_cat, " "_s);
```
Do not add this space before the first argument. You will offset the column of tokens by 1 relative to the start of the argument, which is just bad error reporting.

Kalinovcic commented on src_fun/main.cpp:
```diff
@@ -3,46 +3,110 @@
+        return 0;
+    }
+
+    if (first_flag_arg == "stress_test"_s)
```
Likewise merge -stress_test and -test, and simply check for the -stress flag in the handler.

`fun -test -stress`

Kalinovcic commented on src_fun/main.cpp:
```diff
@@ -3,46 +3,110 @@
 EnterApplicationNamespace
 
 
+static const String TEST_DIRECTORY = "test"_s;
```
After doing the merges below, this will be used in one place and there is no need for this loud constant then.

Kalinovcic commented on src_fun/run.cpp:
```diff
@@ -276,6 +276,16 @@ static bool run_intrinsic(User* user, Unit* unit, byte* storage, Block* block, S
+        get_runtime_parameter("condition"_s, &condition_ptr);
+        if (!(*condition_ptr))
+        {
+            fprintf(stderr, "Assertion failure!\n");
```
Add an `@Incomplete` to add location information to the error. But why is this not implemented in fun? Check comment in testing.cpp

Kalinovcic commented on src_fun/testing.cpp:
```diff
@@ -0,0 +1,483 @@
+static bool run_code_of_test(Test_Case* test)
+{
+    String_Concatenator code_cat = {};
+    add(&code_cat, "test_assert :: (condition: bool) {} intrinsic \"test_assert\";");
```
Instead of crudely prepending the test_assert declaration, you should be adding your declarations to some other preload unit. Then have that be the parent of the test unit.

Then your implementation can comfortably be longer and fully implemented in fun, which removes the unnecessary intrinsic.

This also removes the weird (but currently implicitly satisfied) limitation that code can't start at line 1 without its line numbers getting messed up.

gabriel-marinkovic committed:
changes addressed in PR #2

gabriel-marinkovic commented:
Implemented most of the changes, here is the summary:
- Changed `-test_runner` to `-test_process`; this is only used by test subprocesses
- Most of the test modes are now unified under the `-test` flag, you can add `-stress` to run stress testing, or type in a wildcard/filename to run only those tests.
- All types of tests now show explanations when they fail. This can be noisy when a lot of tests fail, so you can run with `-silent` to silence this.
- Using `-test` with `-inline` runs all specified tests in the same process, one by one. Useful for debugging tests. When a test fails the normal output will give you a command with `-inline` so that you can debug the test.
- All other suggestions and fixes are also implemented.

I also turned on stack trace printing on assertion failures, I have no idea why this was turned off.

Alongside this, there is a larger change to module structuring. This has crudely started now to make loading system modules possible in tests, but will be done properly in a following PR. Modules are now stored in `/modules`, instead of `/src_fun/test/modules`.

PS: Can we delete `src_fun/test` now? The remaining tests have been ported to the testing system and they pass.

gabriel-marinkovic committed:
deleted old unneeded src_fun/test

gabriel-marinkovic committed:
merge conflict resolved (deleting src_fun/test)
